internal/controllers/upgrade_handler.go

```diff
@@ -0,0 +1,157 @@
+	ibgu "github.com/openshift-kni/cluster-group-upgrades-operator/pkg/api/imagebasedgroupupgrades/v1alpha1"
+	"github.com/openshift-kni/oran-o2ims/internal/controllers/utils"
+	siteconfig "github.com/stolostron/siteconfig/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+func (t *provisioningRequestReconcilerTask) IsUpgradeRequested(
+	ctx context.Context, managedClusterName string,
+) (bool, error) {
+	template, err := t.getCrClusterTemplateRef(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get ClusterTemplate: %w", err)
+	}
+
+	if template.Spec.Release == "" {
+		return false, nil
+	}
+
+	managedCluster := &clusterv1.ManagedCluster{}
+	err = t.client.Get(ctx, types.NamespacedName{Name: managedClusterName}, managedCluster)
+	if err != nil {
+		return false, fmt.Errorf("failed to get ManagedCluster: %w", err)
+	}
+
+	templateVersion, err := semver.NewVersion(template.Spec.Release)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse template version: %w", err)
+	}
+	managedClusterVersion, err := semver.NewVersion(managedCluster.GetLabels()["openshiftVersion"])
+	if err != nil {
+		return false, fmt.Errorf("failed to parse ManagedCluster version: %w", err)
+	}
+	cmp := templateVersion.Compare(*managedClusterVersion)
+	if cmp == 1 {
+		return true, nil
+	} else if cmp == -1 {
+		return false, fmt.Errorf("template version (%v) is less then ManagedCluster version (%v)",
+			templateVersion, managedClusterVersion)
+	}
+	return false, nil
+}
+
+func (t *provisioningRequestReconcilerTask) handleUpgrade(
+	ctx context.Context, renderedClusterInstance *siteconfig.ClusterInstance) (ctrl.Result, error) {
+	clusterTemplate, err := t.getCrClusterTemplateRef(ctx)
+	if err != nil {
+		return requeueWithError(fmt.Errorf("failed to get clusterTemplate: %w", err))
+	}
+	ibgu, err := utils.GetIBGUFromUpgradeDefaultsConfigmap(
+		ctx, t.client, clusterTemplate.Spec.Templates.UpgradeDefaults,
+		clusterTemplate.Namespace, utils.UpgradeDefaultsConfigmapKey, renderedClusterInstance.Spec.ClusterName, clusterTemplate.Namespace)
+	if err != nil {
+		return requeueWithError(fmt.Errorf("failed to generate IBGU for cluster: %w", err))
+	}
+	err = t.client.Get(ctx, types.NamespacedName{Name: ibgu.Name, Namespace: ibgu.Namespace}, ibgu)
+	if err != nil && errors.IsNotFound(err) {
+		if err := utils.CreateK8sCR(ctx, t.client, ibgu, t.object, utils.UPDATE); err != nil {
+			return requeueWithError(fmt.Errorf("failed to create IBGU: %w", err))
+		}
+
+		t.logger.InfoContext(
+			ctx,
+			fmt.Sprintf(
+				"Upgrade initiated. Created IBGU %s in the namespace %s",
+				ibgu.GetName(),
+				ibgu.GetNamespace(),
+			),
+		)
+
+		utils.SetStatusCondition(&t.object.Status.Conditions,
+			utils.PRconditionTypes.ClusterProvisioned,
```

Comment:
Based on the doc, it looks like we are considering to add a didicated condition for upgrades?